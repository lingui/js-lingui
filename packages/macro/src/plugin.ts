import type { PluginObj, Visitor, PluginPass } from "@babel/core"
import type * as babelTypes from "@babel/types"
import MacroJSX from "./macroJsx"
import { NodePath } from "@babel/traverse"
import MacroJs from "./macroJs"
import { MACRO_PACKAGE } from "./constants"
import {
  type LinguiConfigNormalized,
  getConfig as loadConfig,
} from "@lingui/conf"
import { Program, Identifier } from "@babel/types"

let config: LinguiConfigNormalized

export type LinguiPluginOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

function reportUnsupportedSyntax(path: NodePath, e: Error) {
  const codeFrameError = path.buildCodeFrameError(
    `Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros.
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues

 Error: ${e.message}`
  )

  // show stack trace where error originally happened
  codeFrameError.stack = e.stack
  throw codeFrameError
}

type LinguiSymbol = "Trans" | "useLingui" | "i18n"

export default function ({
  types: t,
}: {
  types: typeof babelTypes
}): PluginObj {
  function addImport(state: PluginPass, name: LinguiSymbol) {
    const path = state.get(
      "macroImport"
    ) as NodePath<babelTypes.ImportDeclaration>

    const config = state.get("linguiConfig") as LinguiConfigNormalized

    if (!state.get("has_import_" + name)) {
      state.set("has_import_" + name, true)
      const [moduleSource, importName] = config.runtimeConfigModule[name]

      const [newPath] = path.insertAfter(
        t.importDeclaration(
          [
            t.importSpecifier(
              getSymbolIdentifier(state, name),
              t.identifier(importName)
            ),
          ],
          t.stringLiteral(moduleSource)
        )
      )

      path.parentPath.scope.registerDeclaration(newPath)
    }

    return path.parentPath.scope.getBinding(
      getSymbolIdentifier(state, name).name
    )
  }

  function getMacroImports(path: NodePath<Program>) {
    return path.get("body").filter((statement) => {
      return (
        statement.isImportDeclaration() &&
        statement.get("source").node.value === MACRO_PACKAGE
      )
    })
  }

  function getSymbolIdentifier(
    state: PluginPass,
    name: LinguiSymbol
  ): Identifier {
    return state.get("linguiIdentifiers")[name]
  }

  return {
    name: "lingui-macro-plugin",
    visitor: {
      Program: {
        enter(path, state) {
          const macroImports = getMacroImports(path)

          if (!macroImports.length) {
            return
          }

          state.set("macroImport", macroImports[0])

          state.set(
            "linguiConfig",
            getConfig((state.opts as LinguiPluginOpts).linguiConfig)
          )

          state.set("linguiIdentifiers", {
            i18n: path.scope.generateUidIdentifier("i18n"),
            Trans: path.scope.generateUidIdentifier("Trans"),
            useLingui: path.scope.generateUidIdentifier("useLingui"),
          })

          path.traverse(
            {
              JSXElement(path, state) {
                const macro = new MacroJSX(
                  { types: t },
                  {
                    transImportName: getSymbolIdentifier(state, "Trans").name,
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                  }
                )

                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)
                  addImport(state, "Trans").reference(newPath)
                }
              },

              "CallExpression|TaggedTemplateExpression"(
                path: NodePath<
                  | babelTypes.CallExpression
                  | babelTypes.TaggedTemplateExpression
                >,
                state: PluginPass
              ) {
                const macro = new MacroJs(
                  { types: t },
                  {
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                    i18nImportName: getSymbolIdentifier(state, "i18n").name,
                    useLinguiImportName: getSymbolIdentifier(state, "useLingui")
                      .name,
                  }
                )
                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)

                  if (macro.needsUseLinguiImport) {
                    addImport(state, "useLingui").reference(newPath)
                  }

                  if (macro.needsI18nImport) {
                    addImport(state, "i18n").reference(newPath)
                  }
                }
              },
            } as Visitor<PluginPass>,
            state
          )
        },
        exit(path, state) {
          const macroImports = getMacroImports(path)
          macroImports.forEach((path) => path.remove())
        },
      },
    } as Visitor<PluginPass>,
  }
}
