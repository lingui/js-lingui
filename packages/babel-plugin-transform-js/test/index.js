import fs from "fs"
import glob from "glob"
import path from "path"
import { transformFileSync, transform } from "babel-core"

import plugin from "../src/index"

function getTestName(testPath) {
  return path.basename(testPath)
}

describe("babel-plugin-lingui-transform-js", function() {
  const babelOptions = {
    babelrc: false,
    plugins: ["babel-plugin-syntax-jsx", plugin]
  }

  // return function, so we can test exceptions
  const transformCode = code => () => transform(code, babelOptions).code.trim()

  glob.sync(path.join(__dirname, "fixtures/*/")).forEach(testPath => {
    const testName = getTestName(testPath)
    // We're using relative path here to make snapshot testing work
    // across different machines.
    const actualPath = path.relative(
      process.cwd(),
      path.join(testPath, "actual.js")
    )
    const expectedPath = path.join(testPath, "expected.js")

    it(testName, () => {
      const expected =
        fs.existsSync(expectedPath) &&
        fs
          .readFileSync(expectedPath, "utf8")
          .replace(/\r/g, "")
          .trim()

      const actual = transformFileSync(actualPath, babelOptions)
        .code.replace(/\r/g, "")
        .trim()
      expect(actual).toEqual(expected)
    })
  })

  describe("validation", function() {
    describe("plural/select/selectordinal", function() {
      it("value is missing", function() {
        const code = `
        i18n.plural({
          0: "No books",
          1: "1 book",
          other: "# books"
        });`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("offset must be number or string, not variable", function() {
        const code = `
        i18n.plural({
          offset: count,
          0: "No books",
          1: "1 book",
          other: "# books"
        });`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms are missing", function() {
        const plural = `
        i18n.plural({
          value: count
        });`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `
        i18n.plural({
          value: count
        });`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const selectOrdinal = `
        i18n.plural({
          value: count
        });`
        expect(transformCode(selectOrdinal)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms missing fallback", function() {
        const plural = `
        i18n.plural({
          value: count,
          one: "Book"
        });`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const select = `
        i18n.select({
          value: count,
          male: "He"
        });`
        expect(transformCode(select)).toThrowErrorMatchingSnapshot()

        const selectOrdinal = `
        i18n.selectOrdinal({
          value: count,
          one: "st"
        });`
        expect(transformCode(selectOrdinal)).toThrowErrorMatchingSnapshot()
      })

      it("plural forms cannot be variables", function() {
        const code = `
        i18n.plural({
          value: count,
          [one]: "Book"
        });`
        expect(transformCode(code)).toThrowErrorMatchingSnapshot()
      })

      it("plural rules must be valid", function() {
        const plural = `
        i18n.plural({
          value: count,
          one: "Book",
          three: "Invalid",
          other: "Books"
        });`
        expect(transformCode(plural)).toThrowErrorMatchingSnapshot()

        const selectOrdinal = `
        i18n.selectOrdinal({
          value: count,
          one: "st",
          three: "Invalid",
          other: "rd"
        });`
        expect(transformCode(selectOrdinal)).toThrowErrorMatchingSnapshot()
      })
    })
  })

  describe("formats", function() {
    it("value is missing", function() {
      expect(transformCode("i18n.date();")).toThrowErrorMatchingSnapshot()
    })

    it("format must be either string, variable or object with custom format", function() {
      expect(transformCode('i18n.number(value, "currency");')).not.toThrow()
      expect(transformCode("i18n.number(value, currency);")).not.toThrow()
      expect(transformCode("i18n.number(value, { digits: 4 });")).not.toThrow()
      expect(
        transformCode("i18n.number(value, 42);")
      ).toThrowErrorMatchingSnapshot()
    })
  })
})
