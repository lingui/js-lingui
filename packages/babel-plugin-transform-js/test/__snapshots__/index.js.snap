// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-lingui-transform-js formats format must be either string, variable or object with custom format 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Format can be either string for buil-in formats, variable or object for custom defined formats.`;

exports[`babel-plugin-lingui-transform-js formats value is missing 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: The first argument of format function must be a variable.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal offset must be number or string, not variable 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Offset argument cannot be a variable.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms are missing 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Missing plural choices. At least fallback argument 'other' is required.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms are missing 2`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Missing plural choices. At least fallback argument 'other' is required.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms are missing 3`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Missing plural choices. At least fallback argument 'other' is required.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms cannot be variables 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Computed properties aren't allowed.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms missing fallback 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Missing fallback argument 'other'.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms missing fallback 2`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Missing fallback argument 'other'.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural forms missing fallback 3`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Missing fallback argument 'other'.`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural rules must be valid 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Invalid plural rule 'three'. Must be zero, one, two, few, many, other or exact number depending on your source language ('one' and 'other' for English).`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal plural rules must be valid 2`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Invalid plural rule 'three'. Must be zero, one, two, few, many, other or exact number depending on your source language ('one' and 'other' for English).`;

exports[`babel-plugin-lingui-transform-js validation plural/select/selectordinal value is missing 1`] = `packages/babel-plugin-transform-js/test/fixtures/t/actual.js: Value argument is missing.`;
